pkg/cmd/recycle/walker.go

```diff
@@ -0,0 +1,135 @@
+package recycle
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sort"
+)
+
+// Visits a directory tree, depth-first, calling walkFn for every file/directory
+type Walker struct {
+	walkFn filepath.WalkFunc
+
+	// fsuid holds the last fsuid we set
+	fsuid int64
+
+	// lstat is for testing, defaults to os.Lstat
+	lstat func(name string) (fi os.FileInfo, err error)
+
+	// setfsuid is for testing, defaults to syscall.Setfsuid
+	setfsuid func(uid int) (err error)
+
+	// getuid is for testing, defaults to fileinfo.Sys().(*syscall.Stat_t).Uid
+	getuid func(os.FileInfo) (int64, error)
+}
+
+func NewWalker(walkFn filepath.WalkFunc) *Walker {
+	return &Walker{
+		walkFn:   walkFn,
+		fsuid:    -1,
+		lstat:    os.Lstat,
+		setfsuid: setfsuid,
+		getuid:   getuid,
+	}
+}
+
+func (w *Walker) Walk(root string) error {
+	// Lock threads, so our Setfsuid calls always apply to the same thread
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	// TODO: determine original fsuid and switch back?
+
+	// The launching process must have stat permission on the root dir
+	info, err := w.lstat(root)
+	if err != nil {
+		return w.walkFn(root, info, err)
+	}
+	return w.walk(root, info)
+}
+
+func (w *Walker) walk(path string, info os.FileInfo) error {
+	var err error
+
+	// Descend first
+	if info.IsDir() {
+		// switch to the dir's owner
+		err = w.becomeOwner(info)
+		if err != nil {
+			return w.walkFn(path, info, err)
+		}
+
+		// read dir info
+		names, err := readDirNames(path)
+		if err != nil {
+			return w.walkFn(path, info, err)
+		}
+
+		// visit files
+		for _, name := range names {
+			filename := filepath.Join(path, name)
+			fileInfo, err := w.lstat(filename)
+			if err != nil {
+				if err := w.walkFn(filename, fileInfo, err); err != nil {
+					return err
+				}
+			} else {
+				err = w.walk(filename, fileInfo)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	// switch to the file's owner
+	err = w.becomeOwner(info)
+	if err != nil {
+		return w.walkFn(path, info, err)
+	}
+
+	// visit the current file
+	err = w.walkFn(path, info, nil)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (w *Walker) becomeOwner(info os.FileInfo) error {
+	uid, err := w.getuid(info)
+	if err != nil {
+		return err
+	}
+
+	if w.fsuid == uid {
+		return nil
+	}
+
+	fmt.Printf("Becoming %d\n", uid)
+	if err := w.setfsuid(int(uid)); err != nil {
+		return err
+	}
+	w.fsuid = uid
+
+	return nil
+}
+
+// readDirNames reads the directory named by dirname and returns a sorted list of directory entries.
+func readDirNames(dirname string) ([]string, error) {
+	f, err := os.Open(dirname)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	names, err := f.Readdirnames(-1)
```

Comment:
it just returns all children of the current directory
`IsDir()` only returns true for true directories, so we don't recurse links
`Lstat` returns info about links, rather than their targets, so we adjust to permissions of links, not targets